megatron/data/ul2_dataset.py

```diff
@@ -0,0 +1,277 @@
+        denoiser_ratios: Probability of each denoising objective to be selected.
+        denoisers: What type of UL2 denoising objective the other UL2
+              configurations refer to.
+        mean_span_lengths: Mean length for sampling span lengths. Numbers < 1
+              indicate a mean length of the sequence length times that number.
+        mask_ratios: Ratio of masked token in the full sequence.
+        np_rng: Random number genenrator. Note that this rng state should be
+              numpy and not python since python randint is inclusive for
+              the opper bound whereas the numpy one is exclusive.
+        bos_id: start of decoder example id
+        eos_id: end of generation id
+        sentinel_tokens: unique value to be substituted for every replaced span
+    """
+
+    # Denoiser selection
+    denoiser_index = np_rng.choice(np.arange(len(denoisers)), p=denoiser_ratios)
+    denoiser = denoisers[denoiser_index]
+    masked_lm_prob = mask_ratios[denoiser_index]
+
+    assert target_seq_length <= max_seq_length
+
+    # flatten sentences into one list
+    tokens = [token for sentence in sample for token in sentence]
+
+    max_num_tokens = target_seq_length
+    # if is_decoder_only(model_type):
+    #     # Keep space for repeated `extra_id` tokens; not the most data
+    #     # efficient since we calculate this based on the maximum number
+    #     # of possible `extra_id` tokens.
+    #     safe_max_seq_len = math.floor(max_num_tokens / (1 + masked_lm_prob))
+    #     truncated = len(tokens) > safe_max_seq_len
+    #     tokens = tokens[:safe_max_seq_len]
+    # else:
+    # Truncate to `target_sequence_length`.
+    truncated = len(tokens) > max_num_tokens
+    tokens = tokens[:max_num_tokens]
+
+    # Prepend objective token.
+    cls_id = cls_ids.get(denoiser)
+    if cls_id is None:
+        raise ValueError('unknown denoiser')
+    tokens = [cls_id] + tokens
+
+    # Masking.
+    # Ensure we always have at least one prediction.
+    max_predictions_per_seq = max(1.0, masked_lm_prob * len(tokens))
+    mean_ngrams = mean_span_lengths[denoiser_index]
+    if mean_ngrams < 1:
+        # Ensure we always obtain at least one `max_ngrams`.
+        mean_ngrams = max(1, round(len(tokens) * mean_ngrams))
+    max_ngrams = mean_ngrams * 2 - 1
+
+    if denoiser == 'R' or denoiser == 'X':
+        sampling_style = SamplingStyle.NORMAL
+        prefix_lm = False
+    elif denoiser == 'S':
+        sampling_style = SamplingStyle.UNIFORM
+        prefix_lm = True
+    else:
+        raise ValueError('unknown denoiser')
+    (
+        tokens, masked_positions, masked_labels, _, masked_spans,
+    ) = create_masked_lm_predictions(
+        tokens, vocab_id_list, vocab_id_to_token_dict, masked_lm_prob,
+        cls_id, sep_id, mask_id, max_predictions_per_seq, np_rng,
+        max_ngrams=max_ngrams, masking_style="t5",
+        sampling_style=sampling_style, prefix_lm=prefix_lm,
```

Comment:
Once again I trusted the `T5Dataset` with this decision, since the T5 paper also does not mention specifically masking on word boundaries but it is still used as the default. Neither does the UL2 paper mention word splitting, so I thought the same could also be applied here. In short tests with a small dataset, I also did not find `seqio` to split between words. However, this is likely due to the tokenizer.

I personally think your suggestion makes sense and the `T5Dataset` should also have supplied this, although it would make sense to discuss about this. Since real-world prompts will also only split on whole words, this provides an (IMO sensible) inductive bias in the pre-training task.

Reply:
It was not the tokenizer – the `seqio` implementation apparently handles word-splitting, only placing boundaries between words. I couldn't figure out from a quick look where this is implemented, but the tokenizer in my tests did sub-tokenize words, but words were never split.